solidity/random-beacon/contracts/libraries/Relay.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+// TODO: Documentation
+// TODO: Unit tests
+library Relay {
+    struct Request {
+        uint256 id;
+        uint256 startBlock;
+        uint256 groupIndex;
+        bytes previousEntry;
+    }
+
+    struct Data {
+        uint256 requestCount;
+        Request currentRequest;
+        uint256 groupSize;
+        uint256 entrySubmissionBlockStep;
+        uint256 bleedingPeriod;
+    }
+
+    function requestEntry(
+        Data storage self,
+        uint256 groupIndex,
+        bytes calldata previousEntry
+    ) internal {
+        require(!isRequestInProgress(self), "Another relay request in progress");
+
+        self.requestCount++;
+
+        self.currentRequest = Request(
+            self.requestCount,
+            block.number,
+            groupIndex,
+            previousEntry
+        );
+
+        // TODO: Emit event.
+    }
+
+    function submitEntry(
+        Data storage self,
+        bytes calldata entry
+    ) internal {
+        require(isRequestInProgress(self), "No relay request in progress");
+        require(!isRequestTimedOut(self), "Relay request timed out");
+
+        // TODO: Verify entry and submitter eligibility.
```

Comment:
We will need to have access to group data and to the slashing logic. This part can get unwieldy from the library. I would probably start with having this code in the contract but if you have an idea how to set up the interactions and make it gas-effective, that's fine.

Reply:
Yeah, this is still an open topic. I Will play with that and we'll see.